### PR TITLE
Fix autotune agent from passing incorrect configurations and enforce test agent to use input from reference code

### PR DESCRIPTION
**Pull request by shangkunwang01**

### CODE REVIEW

File: MaxKernel/auto_agent/agent_client/run_batch_agent_call.py [hunk 1]
```diff
@@ -89,6 +90,11 @@ def process_problem(
     user_id = "user_0"
     session_id = f"session_{problem_id}_attempt_{attempt}_{int(time.time())}"
 
+    # Add random jitter to avoid SQLite database lock contention
+    jitter = random.uniform(0.1, 2.0)
+    logger.info(f"Sleeping for {jitter:.2f}s (jitter) to avoid DB lock.")
+    time.sleep(jitter)
+
```
Comment (NinaCai) [inline MaxKernel/auto_agent/agent_client/run_batch_agent_call.py:97]: This part is in https://github.com/AI-Hypercomputer/accelerator-agents/pull/46.

File: MaxKernel/auto_agent/subagents/pipeline_agent.py [hunk 2]
```diff
@@ -193,13 +209,42 @@ async def _run_async_impl(
       ),
     )
 
+  def _save_iteration_files(
```
Comment (NinaCai) [inline MaxKernel/auto_agent/subagents/pipeline_agent.py:212]: This code change is in the previous PR too.

File: MaxKernel/auto_agent/subagents/testing/agent.py [hunk 3]
```diff
@@ -619,7 +620,7 @@ async def _run_async_impl(
           "eval_type": "unified_test",
           "code": mock_code_content,
           "timeout": MOCK_EXECUTION_TIMEOUT,
-          "backend_type": "cpu",
+          "backend_type": "tpu",
```
Comment (NinaCai) [inline MaxKernel/auto_agent/subagents/testing/agent.py:623]: Why is it changed to tpu?

File: MaxKernel/auto_agent/subagents/testing/agent.py [hunk 4]
```diff
@@ -33,6 +33,7 @@
 COMPILE_VALIDATION_TIMEOUT = 60 * 1
 MOCK_EXECUTION_TIMEOUT = 60 * 3
 TEST_EXECUTION_TIMEOUT = 60 * 5
+TEST_EXECUTION_POLL_INTERVAL = 20
```
Comment (NinaCai) [inline MaxKernel/auto_agent/subagents/testing/agent.py:36]: Also in previous PR. 

Where is this variable used?

File: MaxKernel/auto_agent/subagents/autotuning/prompts/autotune_prompt.py [hunk 5]
```diff
@@ -3,12 +3,16 @@
...
+3. Create a code_template from the optimized kernel code, replacing the specific parameter values with placeholders enclosed in curly braces (for example, if the parameter is BLOCK_M, use it enclosed in curly braces as the placeholder).
+   To help you build this `code_template`, you must read the reference kernel code located at {base_kernel_path?} and the testing script located at {test_file_path?} to understand the reference computation inputs, outputs, and validation logic.
+   All correctness check and timing logic must be defined inside this `code_template`:
+   - **Reference Computation**: The reference kernel will be automatically written to a file named `base_kernel.py` in the execution directory. Import functions/implementations directly from it (e.g. `from base_kernel import computation as reference_computation, get_inputs`).
+   - **Correctness Check**: In the main block of the template, perform a correctness check comparing the tuned kernel's output against the reference implementation's output (using `jnp.allclose` or `np.testing.assert_allclose` with appropriate tolerances, e.g., atol={atol?}, rtol={rtol?}).
```
Comment (NinaCai) [inline MaxKernel/auto_agent/subagents/autotuning/prompts/autotune_prompt.py]: What is the default values for atol and rtol? I think we should use the same values throughout testing, autotuning, and benchmarking.

File: MaxKernel/auto_agent/subagents/autotuning/prompts/autotune_prompt.py [hunk 6]
```diff
@@ -3,12 +3,16 @@
...
+1. Use `read_file` tool to read the optimized kernel code located at {optimized_kernel_path?}.
+2. Identify the parameters that can be tuned in the kernel (e.g., BLOCK_M, BLOCK_N).
+3. Create a code_template from the optimized kernel code, replacing the specific parameter values with placeholders enclosed in curly braces (for example, if the parameter is BLOCK_M, use it enclosed in curly braces as the placeholder).
+   To help you build this `code_template`, you must read the reference kernel code located at {base_kernel_path?} and the testing script located at {test_file_path?} to understand the reference computation inputs, outputs, and validation logic.
```
Comment (NinaCai) [inline MaxKernel/auto_agent/subagents/autotuning/prompts/autotune_prompt.py:10]: This assumes we have the test file ready. It won't apply to hitl agent. We need to be careful when we do ablation studies(when running without test agent).

File: MaxKernel/auto_agent/subagents/testing/prompts/fix_test_script.py [hunk 7]
```diff
@@ -72,6 +72,8 @@
 2. **DO NOT modify the actual kernel implementations** - only fix the test file
 3. **Keep all test classes and test methods** - just fix syntax/import/structure issues
 4. Focus ONLY on making the test file valid Python code with correct imports and pytest structure
+5. **Numerical Tolerance**: Use the specified tolerances: atol={atol?}, rtol={rtol?}. If they are not specified, default to atol=1e-3, rtol=1e-3.
+6. **Input Generation**: Ensure that if the base kernel file (`{base_kernel_path?}`) defines an input generation function (e.g. `create_inputs`), it is reused/imported and used directly.
```
Comment (NinaCai) [inline MaxKernel/auto_agent/subagents/testing/prompts/fix_test_script.py:76]: atol and rtol should be the same values across different agents and benchmark.

File: MaxKernel/auto_agent/agent_client/run_batch_agent_call.py (hunk 1 above)
Comment (shangkunwang01) [reply MaxKernel/auto_agent/agent_client/run_batch_agent_call.py:97]: Yeah, I did rebase and it showed here. I stack these two pr together. I think once the first one is merge and then rabasing, these overlap will disapper?

File: MaxKernel/auto_agent/subagents/pipeline_agent.py (hunk 2 above)
Comment (shangkunwang01) [reply MaxKernel/auto_agent/subagents/pipeline_agent.py:212]: Same as before.

File: MaxKernel/auto_agent/subagents/testing/agent.py (hunk 3 above)
Comment (shangkunwang01) [reply MaxKernel/auto_agent/subagents/testing/agent.py:623]: If the original code has pallas kernel, mock test can fail.

File: MaxKernel/auto_agent/subagents/testing/agent.py (hunk 4 above)
Comment (shangkunwang01) [reply MaxKernel/auto_agent/subagents/testing/agent.py:36]: Same as above.

File: MaxKernel/auto_agent/subagents/autotuning/prompts/autotune_prompt.py (hunk 5 above)
Comment (shangkunwang01) [reply MaxKernel/auto_agent/subagents/autotuning/prompts/autotune_prompt.py]: In my experiment it is set as 1e-2. autotune will look at that value as well as the test file to use the same tol.

File: MaxKernel/auto_agent/subagents/autotuning/prompts/autotune_prompt.py (hunk 6 above)
Comment (shangkunwang01) [reply MaxKernel/auto_agent/subagents/autotuning/prompts/autotune_prompt.py:10]: Without test file, it will rely only on the reference code.

File: MaxKernel/auto_agent/subagents/testing/prompts/fix_test_script.py (hunk 7 above)
Comment (shangkunwang01) [reply MaxKernel/auto_agent/subagents/testing/prompts/fix_test_script.py:76]: atol and rtol is set as 1e-2 for all agent as they are the state variable hardcoded in pipeline agent. The idea of using same tol value is perhaps not correct because some task may inherently be difficult to achieve the same level of accuracy. For example in minibench, each task is paired with its own specified atol and rtol.

File: MaxKernel/auto_agent/agent_client/run_batch_agent_call.py (hunk 1 above)
Comment (NinaCai) [reply MaxKernel/auto_agent/agent_client/run_batch_agent_call.py:97]: You don't need to rebase the code from other PRs. In fact, this will make merging more complex. If you are to revert the rebase, let's do this.

File: MaxKernel/auto_agent/subagents/autotuning/prompts/autotune_prompt.py (hunk 6 above)
Comment (NinaCai) [reply MaxKernel/auto_agent/subagents/autotuning/prompts/autotune_prompt.py:10]: How does it generate correctness test without referring to test file and only on the reference code?

File: MaxKernel/auto_agent/subagents/testing/prompts/fix_test_script.py (hunk 7 above)
Comment (NinaCai) [reply MaxKernel/auto_agent/subagents/testing/prompts/fix_test_script.py:76]: For jaxbench we are using 1e-1, and for autotune and test we are using 1e-2. This is a more strict rule for agent than for benchmark, right? Meaning, we filter out many candidates from autotune but they are actually valid during benchmark?

File: MaxKernel/auto_agent/subagents/pipeline_agent.py (hunk 2 above)
Comment (NinaCai) [reply MaxKernel/auto_agent/subagents/pipeline_agent.py:212]: Let's remove the changes from the other PR.

File: MaxKernel/auto_agent/subagents/testing/prompts/fix_test_script.py (hunk 7 above)
Comment (shangkunwang01) [reply MaxKernel/auto_agent/subagents/testing/prompts/fix_test_script.py:76]: I observed that agent sometimes realize the tol is too strict for that precision and then set a looser one. For example, in the gqa attention case, I see the agent change tol from 1e-2 to 5e-2.
I would suggest using a 1e-2 tol for the agent and ideally we should use 1e-2 for benchmark as well (I will actually try this). But give the benchmark sometimes is not that reasonable, I set a loose 1e-1 tol in benchmark now.

File: MaxKernel/auto_agent/subagents/autotuning/prompts/autotune_prompt.py (hunk 6 above)
Comment (shangkunwang01) [reply MaxKernel/auto_agent/subagents/autotuning/prompts/autotune_prompt.py:10]: I put a reference to the test file here is just to help autotune to generate test. Without this prompt, I think it will come up with the test just like test agent. 
When you adapt this to hitl, you may try if the agent can generate good test by removing this from prompt.

**Review by NinaCai — APPROVED** [review]
This PR rebased https://github.com/AI-Hypercomputer/accelerator-agents/pull/46.

After merging this PR, let's close PR 46. Let's make separate PRs without rebasing each other next time.
